### PR TITLE
Handle concurrent downloads of same file better

### PR DESCRIPTION
**Pull request by jcpetruzza**
Previously, in order to download a file `foo/bar.json`, first `foo/bar.json.tmp` would be downloaded and then it would get renamed to `forr/bar.json`. The problem with this approach is that it doesn't handle concurrent downloads well.

For example, if two processes are trying to download this file at the same time, the second one to finish will likely fail with an error of the form:

```
foo/bar.json.tmp: renameFile:renamePath:rename: does not exist (No such file or directory)
```

This is because the faster process already renamed that file.

To fix this, we now download the file to `foo/verifiedDownloadZZZZ/bar.json` where `ZZZZ` is guaranteed to be unique. Because it is created on the same filesystem, renaming is still possible.

In order to use `Path.IO.withTempDir`, we need to introduce a dependency on `rio-orphans`, that provides the required `MonadMask (RIO env)` instance.

Note that this problem currently manifests on `stack` when using global hints. In particular, we observe it on CI when simultaneous builds run on the same box (https://github.com/tweag/rules_haskell/issues/1167)

### CODE REVIEW

**Comment by snoyberg** [comment]
I'd rather not use `rio-orphans` for this; explicit usage of `unliftio` can help. However, I'm confused about the usage of `withTempDir`. Why not use `withTempFile`?

**Comment by jcpetruzza** [comment]
Pushed a new version that avoids `rio-orphans` (wasn't aware of `UnlitIO.Temporary`).

> Why not use withTempFile?

If I understand correctly, `withTempFile` is based on `System.IO.openTempfile`, and its docs say that the temp file is created with permissions so that only the current user can read/write. So I think we'd be changing the semantics of `download` if we did. This way we create the file with exactly the same permissions as before.

**Comment by snoyberg** [comment]
I'm not sure if those semantic differences are important or intended. @qrilka can you weigh in on the file vs directory discussion here?

**Comment by qrilka** [comment]
Sorry for a late reply. I think @snoyberg is right and we should use `withTempFile`, the details of what `System.IO.openTempfile` don't seem to change the way this function works actually `O_CREAT | O_EXCL` used in that function should exclude a possible conflict of 2 concurrent accesses to the same file and that's exactly what seems to be required here. And it's not quite about permissions.

**Comment by jcpetruzza** [comment]
Sorry I couldn't look into this earlier. I pushed a change that uses `withTempFile` as suggested. I agree that this is more secure. However, I still think that as it is, we'd be slightly changing the behavior of `verifiedDownload`:
  - Before the change, on POSIX systems, the downloaded file is created with the default file permissions for the user, as controlled by the `umask`.  
  - After the change, the downloaded file will have r/w permission for the owner, and no access any other user.

This is because `withTempFile` uses `System.IO.openTempFile` that is defined as follows, where `0o600` is the mode:

```haskell
openTempFile tmp_dir template
    = openTempFile' "openTempFile" tmp_dir template False 0o600
```

The only alternative I can think of is to use a copy of `withTempFile` based on `System.IO.openTempFileWithDefaultPermissions` instead.

**Comment by jcpetruzza** [comment]
The latest version uses a `withTempFileDefaultPermissions` function

File: src/Network/HTTP/Download/Verified.hs [hunk 1]
```diff
@@ -322,3 +324,21 @@ verifiedDownload DownloadRequest{..} destpath progressSink = do
...
+             -> String   -- ^ File name template. See 'openTempFile'.
+             -> (FilePath -> Handle -> m a) -- ^ Callback that can use the file.
+             -> m a
+withTempFileWithDefaultPermissions tmpDir template action =
+  bracket
+    (liftIO (openTempFileWithDefaultPermissions tmpDir template))
+    (\(name, handle') -> liftIO (hClose handle' >> ignoringIOErrors (removeFile name)))
+    (uncurry action)
+  where
+    ignoringIOErrors = liftM (const ()) . tryIO -- yes, it's just void, but for pre-AMP GHCs
```
Comment (snoyberg) [inline src/Network/HTTP/Download/Verified.hs]: Using `liftM` here triggers a CI failure:

https://dev.azure.com/commercialhaskell/http-download/_build/results?buildId=2622&view=logs&jobId=4556e1b3-9539-56cd-971d-ac2ca862cf24&j=4556e1b3-9539-56cd-971d-ac2ca862cf24&t=5e9614ff-3cd8-56ab-009b-b533691987fd

Can you switch it to `fmap`? Actually, given that we don't support pre-AMP GHCs, using `void . tryIO` is probably just fine.

**Review by snoyberg — CHANGES_REQUESTED** [review]

File: src/Network/HTTP/Download/Verified.hs (hunk 1 above)
Comment (jcpetruzza) [reply src/Network/HTTP/Download/Verified.hs]: Changed to use `void`

**Review by snoyberg — APPROVED** [review]
Thanks!

**Pull request merged**